front/src/applications/operationalStudies/components/MacroEditor/OsrdNgeSync.tsx

```diff
@@ -0,0 +1,897 @@
+        continue; // track offset, handled by creating an empty node
+      }
+
+      // Avoid including the same query twice in the search payload
+      const key = JSON.stringify(query);
+      if (pathItemSet.has(key)) {
+        continue;
+      }
+
+      pathItemSet.add(key);
+      pathItemQueries.push(query);
+    }
+
+    if (pathItemQueries.length === 0) {
+      return null;
+    }
+
+    return {
+      object: 'operationalpoint',
+      query: ['and', ['=', ['infra_id'], this.scenario.infra_id], ['or', ...pathItemQueries]],
+    };
+  }
+
+  /**
+   * Cast a node into NGE format.
+   */
+  private castNodeToNge(node: MacroNodeResponse): NetzgrafikDto['nodes'][0] {
+    const labelsList = Array.from(this.labels);
+    return {
+      id: node.id,
+      betriebspunktName: node.trigram || '',
+      fullName: node.full_name || '',
+      positionX: node.position_x,
+      positionY: node.position_y,
+      ports: [],
+      transitions: [],
+      connections: [],
+      resourceId: this.ngeResource.id,
+      perronkanten: 10,
+      connectionTime: node.connection_time,
+      trainrunCategoryHaltezeiten: TRAINRUN_CATEGORY_HALTEZEITEN,
+      symmetryAxis: 0,
+      warnings: [],
+      labelIds: (node.labels || [])?.map((l) => labelsList.findIndex((e) => e === l)),
+    };
+  }
+
+  /**
+   * Cast a NGE node to a node.
+   */
+  static castNgeNode(
+    node: NetzgrafikDto['nodes'][0],
+    labels: NetzgrafikDto['labels']
+  ): Omit<MacroNodeResponse, 'path_item_key'> {
+    return {
+      id: node.id,
+      trigram: node.betriebspunktName,
+      full_name: node.fullName,
+      connection_time: node.connectionTime,
+      position_x: node.positionX,
+      position_y: node.positionY,
+      labels: node.labelIds
+        .map((id) => {
+          const ngeLabel = labels.find((e) => e.id === id);
+          if (ngeLabel) return ngeLabel.label;
+          return null;
+        })
+        .filter((n) => n !== null) as string[],
+    };
+  }
+
+  /**
+   * Get nodes of the scenario that are saved in the DB.
+   */
+  private async apiGetSavedNodes(): Promise<MacroNodeResponse[]> {
+    const pageSize = 100;
+    let page = 1;
+    let reachEnd = false;
+    const result: MacroNodeResponse[] = [];
+    while (!reachEnd) {
+      const promise = this.dispatch(
+        osrdEditoastApi.endpoints.getProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNodes.initiate(
+          {
+            projectId: this.scenario.project.id,
+            studyId: this.scenario.study_id,
+            scenarioId: this.scenario.id,
+            pageSize,
+            page,
+          }
+        )
+      );
+      // need to unsubscribe on get call to avoid cache issue
+      promise.unsubscribe();
+      const { data } = await promise;
+      if (data) result.push(...data.results);
+      if (!data?.next) reachEnd = true;
+      page += 1;
+    }
+    return result;
+  }
+
+  private async apiCreateNode(node: Omit<MacroNodeResponse, 'id'>) {
```

Comment:
This is something used for the NGE → OSRD sync. I don't think it's a great idea to mix up both sides in the same file.

Reply:
To make the sync between OSRD, NGE & API/DB, there is state in the middle, which tracks nodes for their IDs, `path_item_key` and if they are saved, but also labels. So when you do osrd->nge you need to modify it, and vice-versa.

I agree that having a clear code frontier between the sync way is good. If you have implementation ideas, I'm opened to discuss.

Reply:
My original plan for this was to keep `osrdToNge.ts` and `ngeToOsrd.ts` as-is (do not introduce a class), and add a shared "context" type passed around to their functions. Something like:

```ts
type NgeContext = {
  nodesByPathKey: Map<String, …>;
  // etc
};
```